_posts/2023-10-23-unidirectional-topic-operator.md

```diff
@@ -0,0 +1,312 @@
+While the former is not supported by Kafka itself, support for the latter could be added in the future.
+
+In addition to be able to pause reconciliations, the operator now also provides a way to unmanage a Kafka topic using annotations.
+If the `KafkaTopic` is paused and we delete it, then it will be also deleted in Kafka.
+If the `KafkaTopic` is unmanaged and we delete it, then only the resource will be garbage collected by Kubernetes, but the topic will still exists in Kafka.
+
+In terms of requirements, the UTO assumes the following grants:
+
+- describe all topics
+- describe all topics configs
+- create topics
+- create partitions
+- delete topics
+- list partition reassignments
+- describe broker configs
+
+#### Scalability
+
+The BTO is limited in terms of scalability, as it only operates on one topic at a time, and it has to update a persistent store containing topic metadata.
+Instead, the UTO does not store topic metadata and it aims to be scalable in terms of the number of topics that it can operate on.
+The following line graph confirms the BTO scalability issue, while the UTO scales almost linearly.
+
+<figure>
+    <img src="/assets/images/posts/2023-10-23-uto-max-recon-time.png" height=350>
+    <figcaption><small>
+        Fig 2. Line graph comparing BTO and UTO max reconciliation times.<br/>
+        We used a 3-nodes cluster with default configurations, running on a local Minikube instance with 10 cores and 28 GB of memory.<br/>
+        The test driver application was running on the same machine, and the time spent on the event queue was not included.
+    </small></figcaption>
+</figure>
+
+When running Kafka operations, the UTO makes use of the request batching supported by the Kafka Admin client to get higher throughput for metadata operations.
+All `KafkaTopic` events are queued when received, and then processed in batches by a number of controller threads (currently, only a single thread is supported).
+If more than one event related to a single topic resource are found in the batch building, they are put back in the queue to be processed with the next batch.
+
+You can tune the batching mechanism by setting `STRIMZI_MAX_QUEUE_SIZE` (default: 1024), `STRIMZI_MAX_BATCH_SIZE` (default: 100), and `MAX_BATCH_LINGER_MS` (default: 100).
+If you exceed the configured max queue size, the UTO will print an error and then shutdown (Kubernetes will take care of restarting the pod).
+In that case, you can simply raise the max queues size to avoid the periodic operator restarts.
+
+In another UTO test, we set max queue size to `MAX_INT`, max batch size to 200, batch linger ms to 500 and reconciliation interval to 10 seconds.
+With these settings, we were able to achieve a max reconciliation time of 2.5 seconds with 10k concurrent topic events.
+This translates to 3325 new `KafkaTopic` custom resources to become ready in about 2 minutes and 40 seconds.
```

Comment:
What exactly does this mean? Creating 10k Kubernetes resources normally takes some time. So are they pre-created when you start the operator? This would deserve explanation because users might try it ;-).

Reply:
Will try to improve the description, but topics are not pre-created. This [test driver](https://github.com/fvaleri/examples/blob/main/misc/strimzi-load/src/main/java/it/fvaleri/example/Main.java#L77) first creates 5k topics and then you have 2.5k random update attempts and 2.5k random delete attempts. This test ends up with around 3.3k topics and I'm reporting the readiness time (2.4 min).

Reply:
Well, that does not sound like 10k concurrent events. If you create 5k new topics, how come only 3325 are created in the 2:40 interval? How quickly are full 5.000 topics created? Is the 2:40 interval really caused by the UTO or by the topic creation?

----

I get that a test like this can be helpful _internally_ to ensure the UTO has the expected performance after some changes. But I don't think this can be presented as a scalability numbers.